applicationset/generators/scm_provider.go

```diff
@@ -60,6 +65,26 @@ func (g *SCMProviderGenerator) GetTemplate(appSetGenerator *argoprojiov1alpha1.A
 	return &appSetGenerator.SCMProvider.Template
 }
 
+func ScmProviderAllowed(applicationSetInfo *argoprojiov1alpha1.ApplicationSet, url string, allowedScmProviders []string) bool {
+	if url == "" || len(allowedScmProviders) == 0 {
+		return true
+	}
+
+	for _, allowedScmProvider := range allowedScmProviders {
+		if strings.HasPrefix(url, allowedScmProvider) {
```

Comment:
Should we add a check if the `allowedScmProvider` is a URL as well? 

If user sets the value as `www.example.com` in `allowedScmProviders` and the url is `https://www.example.com`, it will fail the check. It would be more of an additional check.

Reply:
We already had this discussion here: https://github.com/argoproj/argo-cd/pull/14286#discussion_r1257354821

According to me, as this is the base url (ie: https://gitlab.example.com/) we should keep it simple by not considering the protocol. I am using `HasPrefix` but I would be even in favor of using `==` so that the url used is exactly the same as `allowedScmProvider`. 

What do you think  @ishitasequeira and @crenshaw-dev

Reply:
Yeah I like simple ==, as long as it's documented. URL comparison gets messy fast.

Reply:
Yeah.. == works..

Reply:
Fixed

Reply:
We may have an issue using == in the e2e

httptest assigns random ports. ie: http://127.0.0.1:40499

When I was using `HasPrefix` it was working as `http://127.0.0.1` is a prefix of `http://127.0.0.1:40499`
(`ARGOCD_APPLICATIONSET_CONTROLLER_ALLOWED_SCM_PROVIDERS=http://localhost,http://127.0.0.1`)

So either I can come back to HasPrefix or we can find a way to set the port number in httptest.

Currently investigating solution 2

Reply:
I found a way to fix the port number:

```golang
       // create a listener with the desired port.
	l, err := net.Listen("tcp", "127.0.0.1:8341")
	if err != nil {
		t.Error(fmt.Errorf("Unable to start server %w", err))
	}

	ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		githubSCMMockHandler(t)(w, r)
	}))

	// NewUnstartedServer creates a listener. Close that listener and replace
	// with the one we created.
	ts.Listener.Close()
	ts.Listener = l

	// Start the server.
	ts.Start()
	// Stop the server on return from the function.
	defer ts.Close()

       // Test
```

Then I can set:

```
ARGOCD_APPLICATIONSET_CONTROLLER_ALLOWED_SCM_PROVIDERS=http://127.0.0.1:8341 
```

Reply:
I would recommend to switch to this and hoping we will not have conflict with port 8341 :D

Reply:
I think that's reasonable